docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,391 @@
+# ADR 038: KVStore state listening
+
+## Changelog
+
+- 11/23/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.
+
+## Context
+
+Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
+In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.
+
+## Decision
+
+We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to specific state changes in underlying KVStores.
+We will also introduce two approaches for exposing the data to external consumers: writing to files and writing to a gRPC stream.
+
+### Listening interface
+In a new file- `store/types/listening.go`- we will create a `Listening` interface for streaming out an allowed subset of state changes from a KVStore.
+The interface can be backed by a simple wrapper around any underlying `io.Writer`.
+
+```go
+// Listening interface comprises the methods needed to filter and stream data out from a KVStore
+type Listening interface {
+	io.Writer // the underlying io.Writer
+	Allowed(op Operation, key []byte) bool // method used to check if the Listener is allowed to listen to a specific state change
+	TraceContext() TraceContext // method to access this Listener's TraceContext
+}
+```
+
+We will lift some of the private types currently being used by the `tracekv.Store` into public types at a new location- `store/types/tracing.go`- for use in the `Listening` interface
+```go
+type Operation string
+
+const (
+	WriteOp     Operation = "write"
+	ReadOp      Operation = "read"
+	DeleteOp    Operation = "delete"
+	IterKeyOp   Operation = "iterKey"
+	IterValueOp Operation = "iterValue"
```

Comment:
I see no use case for listening to anything besides writes and deletions. Maybe there's is one but it isn't explained and seems out or scope of this ADR

Reply:
I think there could be value to being able to watch the stateless operations, e.g. listening to `Get` operations to watch a node's read access. Happy to either expand on in the ADR that or adjust to focus on writes and deletes.

Reply:
I mean maybe there's value in that but it's not actually watching for state changes so seems out of scope.